// @monaco-editor/react/lib/cjs/DiffEditor/DiffEditor.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var monaco = require('@monaco-editor/loader');
var React = require('react');
var PropTypes = require('prop-types');
var index$2 = require('../MonacoContainer/index.js');
var index = require('../hooks/useMount/index.js');
var index$1 = require('../hooks/useUpdate/index.js');
var index$3 = require('../utils/index.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var monaco__default = /*#__PURE__*/_interopDefaultLegacy(monaco);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

function DiffEditor({
  original,
  modified,
  language,
  originalLanguage,
  modifiedLanguage,
  editorDidMount,
  theme,
  width,
  height,
  loading,
  options,
  className,
  wrapperClassName
}) {
  const [isEditorReady, setIsEditorReady] = React.useState(false);
  const [isMonacoMounting, setIsMonacoMounting] = React.useState(true);
  const editorRef = React.useRef(null);
  const monacoRef = React.useRef(null);
  const containerRef = React.useRef(null);
  const editorDidMountRef = React.useRef(editorDidMount);
  index['default'](() => {
    const cancelable = monaco__default['default'].init();
    cancelable.then(monaco => (monacoRef.current = monaco) && setIsMonacoMounting(false)).catch(error => (error === null || error === void 0 ? void 0 : error.type) !== 'cancelation' && console.error('Monaco initialization: error:', error));
    return () => editorRef.current ? disposeEditor() : cancelable.cancel();
  });
  index$1['default'](() => {
    editorRef.current.getModel().modified.setValue(modified);
  }, [modified], isEditorReady);
  index$1['default'](() => {
    editorRef.current.getModel().original.setValue(original);
  }, [original], isEditorReady);
  index$1['default'](() => {
    const {
      original,
      modified
    } = editorRef.current.getModel();
    monacoRef.current.editor.setModelLanguage(original, originalLanguage || language);
    monacoRef.current.editor.setModelLanguage(modified, modifiedLanguage || language);
  }, [language, originalLanguage, modifiedLanguage], isEditorReady);
  index$1['default'](() => {
    monacoRef.current.editor.setTheme(theme);
  }, [theme], isEditorReady);
  index$1['default'](() => {
    editorRef.current.updateOptions(options);
  }, [options], isEditorReady);
  const setModels = React.useCallback(() => {
    const originalModel = monacoRef.current.editor.createModel(original, originalLanguage || language);
    const modifiedModel = monacoRef.current.editor.createModel(modified, modifiedLanguage || language);
    editorRef.current.setModel({
      original: originalModel,
      modified: modifiedModel
    });
  }, [language, modified, modifiedLanguage, original, originalLanguage]);
  const createEditor = React.useCallback(() => {
    editorRef.current = monacoRef.current.editor.createDiffEditor(containerRef.current, {
      automaticLayout: true,
      ...options
    });
    setModels();
    monacoRef.current.editor.setTheme(theme);
    setIsEditorReady(true);
  }, [options, theme, setModels]);
  React.useEffect(() => {
    if (isEditorReady) {
      const {
        original,
        modified
      } = editorRef.current.getModel();
      editorDidMountRef.current(modified.getValue.bind(modified), original.getValue.bind(original), editorRef.current);
    }
  }, [isEditorReady]);
  React.useEffect(() => {
    !isMonacoMounting && !isEditorReady && createEditor();
  }, [isMonacoMounting, isEditorReady, createEditor]);

  const disposeEditor = () => editorRef.current.dispose();

  return /*#__PURE__*/React__default['default'].createElement(index$2['default'], {
    width: width,
    height: height,
    isEditorReady: isEditorReady,
    loading: loading,
    _ref: containerRef,
    className: className,
    wrapperClassName: wrapperClassName
  });
}

DiffEditor.propTypes = {
  original: PropTypes__default['default'].string,
  modified: PropTypes__default['default'].string,
  language: PropTypes__default['default'].string,
  originalLanguage: PropTypes__default['default'].string,
  modifiedLanguage: PropTypes__default['default'].string,
  editorDidMount: PropTypes__default['default'].func,
  theme: PropTypes__default['default'].string,
  width: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string]),
  height: PropTypes__default['default'].oneOfType([PropTypes__default['default'].number, PropTypes__default['default'].string]),
  loading: PropTypes__default['default'].oneOfType([PropTypes__default['default'].element, PropTypes__default['default'].string]),
  options: PropTypes__default['default'].object,
  className: PropTypes__default['default'].string,
  wrapperClassName: PropTypes__default['default'].string
};
DiffEditor.defaultProps = {
  editorDidMount: index$3.noop,
  theme: 'light',
  width: '100%',
  height: '100%',
  loading: 'Loading...',
  options: {}
};

exports.default = DiffEditor;
